### ISSUE TITLE
IllegalReferenceCountException error with the following benchmark

### ISSUE THREAD

**Issue by mossuchida** [Hyperfoil/Hyperfoil#108]
I have run the attached. It runs & gets result (You can see 2nd line, "Decision is...'), but it is failing with following error.   Please have a look:

```
13:50:28,962 INFO  (vert.x-eventloop-thread-0) [i.h.c.ControllerPhase] 001F main changing status STARTING to RUNNING
13:50:29,311 INFO  (nioEventLoopGroup-2-1) [i.h.c.s.LogStep] Decision is {"decision":"Permit","obligations":[{"actions":["GET /perfguild"],"crns":[""],"decision":"Permit","max-age":86400,"obligationId":"bf69c79fbe636dc8"}]}
13:50:29,733 WARN  (nioEventLoopGroup-2-1) [i.n.c.DelegatingChannelPromiseNotifier] Failed to mark a promise as success because it has failed already: DefaultChannelPromise@509b6506(failure: io.netty.handler.codec.EncoderException: io.netty.util.IllegalReferenceCountException: refCnt: 0, decrement: 1), unnotified cause: io.netty.handler.codec.EncoderException: io.netty.util.IllegalReferenceCountException: refCnt: 0, decrement: 1
    at io.netty.handler.codec.MessageToMessageEncoder.write(MessageToMessageEncoder.java:107)
    at io.netty.channel.CombinedChannelDuplexHandler.write(CombinedChannelDuplexHandler.java:348)
    at io.netty.channel.AbstractChannelHandlerContext.invokeWrite0(AbstractChannelHandlerContext.java:716)
    at io.netty.channel.AbstractChannelHandlerContext.invokeWriteAndFlush(AbstractChannelHandlerContext.java:763)
    at io.netty.channel.AbstractChannelHandlerContext.write(AbstractChannelHandlerContext.java:789)
    at io.netty.channel.AbstractChannelHandlerContext.writeAndFlush(AbstractChannelHandlerContext.java:757)
    at io.hyperfoil.core.client.netty.Http1xConnection.request(Http1xConnection.java:234)
    at io.hyperfoil.core.client.netty.HttpConnectionPoolImpl.request(HttpConnectionPoolImpl.java:91)
    at io.hyperfoil.core.steps.HttpRequestStep.invoke(HttpRequestStep.java:135)
    at io.hyperfoil.api.session.SequenceInstance.progress(SequenceInstance.java:28)
    at io.hyperfoil.core.session.SessionImpl.runSession(SessionImpl.java:278)
    at io.hyperfoil.core.session.SessionImpl.call(SessionImpl.java:231)
    at io.hyperfoil.core.session.SessionImpl.call(SessionImpl.java:32)
    at io.netty.util.concurrent.PromiseTask.run(PromiseTask.java:73)
    at io.netty.util.concurrent.AbstractEventExecutor.safeExecute(AbstractEventExecutor.java:163)
    at io.netty.util.concurrent.SingleThreadEventExecutor.runAllTasks(SingleThreadEventExecutor.java:510)
    at io.netty.channel.nio.NioEventLoop.run(NioEventLoop.java:518)
    at io.netty.util.concurrent.SingleThreadEventExecutor$6.run(SingleThreadEventExecutor.java:1044)
    at io.netty.util.internal.ThreadExecutorMap$2.run(ThreadExecutorMap.java:74)
    at io.netty.util.concurrent.FastThreadLocalRunnable.run(FastThreadLocalRunnable.java:30)
    at java.lang.Thread.run(Thread.java:748)
Caused by: io.netty.util.IllegalReferenceCountException: refCnt: 0, decrement: 1
    at io.netty.util.internal.ReferenceCountUpdater.toLiveRealRefCnt(ReferenceCountUpdater.java:74)
    at io.netty.util.internal.ReferenceCountUpdater.release(ReferenceCountUpdater.java:138)
    at io.netty.buffer.AbstractReferenceCountedByteBuf.release(AbstractReferenceCountedByteBuf.java:100)
    at io.netty.handler.codec.http.DefaultFullHttpRequest.release(DefaultFullHttpRequest.java:102)
    at io.netty.util.ReferenceCountUtil.release(ReferenceCountUtil.java:88)
    at io.netty.handler.codec.MessageToMessageEncoder.write(MessageToMessageEncoder.java:91)
    ... 20 more

```
[log.tar.gz](https://github.com/Hyperfoil/Hyperfoil/files/4668671/log.tar.gz)

[test.yaml.zip](https://github.com/Hyperfoil/Hyperfoil/files/4668696/test.yaml.zip)

**Comment by rvansa** [comment]
Looks like a problem in `ConstantBytesGenerator`, in the `apply` we should increment refcount on the returned buffer (`.retain()`).

**Comment by mossuchida** [comment]
The error is gone, but now I am getting 400 error for consequent calls

```
17:48:28,770 INFO  (nioEventLoopGroup-2-1) [i.h.c.s.LogStep] Decision is {"decision":"Permit","obligations":[{"actions":["GET /perfguild"],"crns":[""],"decision":"Permit","max-age":86400,"obligationId":"bf69c79fbe636dc8"}]}
17:48:30,987 WARN  (nioEventLoopGroup-2-1) [i.h.c.h.RangeStatusValidator] #0 Sequence authorization(0)(3/4), connection Http1xConnection{/9.26.62.76:19106 -> cp-console.cs-9ad9865e1d8eb33160762a5e0740d3a7-0000.us-east.containers.appdomain.cloud/52.117.83.74:443, size=1} received invalid status 400
17:48:34,643 WARN  (nioEventLoopGroup-2-1) [i.h.c.h.RangeStatusValidator] #0 Sequence authorization(0)(3/4), connection Http1xConnection{/9.26.62.76:19114 -> cp-console.cs-9ad9865e1d8eb33160762a5e0740d3a7-0000.us-east.containers.appdomain.cloud/52.117.83.74:443, size=1} received invalid status 400
17:48:36,350 WARN  (nioEventLoopGroup-2-1) [i.h.c.h.RangeStatusValidator] #0 Sequence authorization(0)(3/4), connection Http1xConnection{/9.26.62.76:19118 -> cp-console.cs-9ad9865e1d8eb33160762a5e0740d3a7-0000.us-east.containers.appdomain.cloud/52.117.83.74:443, size=1} received invalid status 400
17:48:37,912 WARN  (nioEventLoopGroup-2-1) [i.h.c.h.RangeStatusValidator] #0 Sequence authorization(0)(3/4), connection Http1xConnection{/9.26.62.76:19120 -> cp-console.cs-9ad9865e1d8eb33160762a5e0740d3a7-0000.us-east.containers.appdomain.cloud/52.117.83.74:443, size=1} received invalid status 400
17:48:38,546 WARN  (nioEventLoopGroup-2-1) [i.h.c.h.RangeStatusValidator] #0 Sequence authorization(0)(3/
```
It looks like payload is wrong
```
[i.h.c.s.SessionImpl] #0 authDecision <- {"error":"Invalid request payload"}
```

I tried to see the traffic using wireshark, but everything is encrypted, I cannot see the content. It seems like I can add logging in `i.h.c.s.HttpRequestStep` to see the header & body, but I am not sure which variables contain headers & request body (& URL).  Would you let me know which variables hold these info that I can print out the log?

[hyperfoil.local.log.zip](https://github.com/Hyperfoil/Hyperfoil/files/4685182/hyperfoil.local.log.zip)

**Comment by rvansa** [comment]
So it seems that the non-allocating version doesn't work. I've over-optimized that not realizing that the code that sends the data will mutate the `ByteBuf`. I am afraid we need to go for this (allocating) implementation, and later implement pooling of the buffers wrapping constant array:
```
public class ConstantBytesGenerator implements SerializableBiFunction<Session, Connection, ByteBuf> {
   private final byte[] bytes;

   public ConstantBytesGenerator(byte[] bytes) {
      this.bytes = bytes;
   }

   @Override
   public ByteBuf apply(Session session, Connection connection) {
      // TODO: implement pooling of wrapping buffers.
      // We cannot cache this buffer since despite the wrapped array is constant the code
      // writing the buffer to socket would mutate the ByteBuf (move readerIndex/writerIndex, decrement refCnt...) 
      return Unpooled.wrappedBuffer(bytes);
   }
}
```

**Comment by rvansa** [comment]
I was able to verify the behaviour you've described by just raising number of repetitions in one of the existing tests. Fixed in master.

**Comment by mossuchida** [comment]
Thanks! I just don't know how you can figure out the issue & able to fix like this.  Looking at the trace & error, it was not clear to me. I will keep learning!

**Comment by mossuchida** [comment]
Verified it works.

**Comment by rvansa** [comment]
The exception mentions reference counting on ByteBuf objects, and it's in the send part. So I've traced back the source of the buffer. Seeing that it's sort of 'static' object (not created in each invocation but only when the generator - part of the scenario - is constructed/deserialized, and knowing that the counter goes below zero (from the exception) I realized that it is a problem with its reuse.
First I've tried to fix it with incrementing the counter (`retain()`) before each use. However after seeing that your server is not getting the expected input (but empty body) I figured out that the buffer is used in mutating way, rendering it 'consumed'. I quickly pondered about resetting the `readerIndex` (unconsume it) but then I realized that it would not work, since the buffer is used in concurrent threads, and even within a single thread it might be just 'queued' for consumption later (e.g. if the TCP socket is congested). So I had to go for the most common create&forget approach (despite that produces some garbage objects).
